azkaban-exec-server/src/main/java/azkaban/dag/DagBuilder.java

```diff
@@ -69,13 +70,100 @@ public NodeBuilder createNode(final String name, final NodeProcessor nodeProcess
    * @return the Dag reflecting the current state of the DagBuilder
    */
   public Dag build() {
+    checkCircularDependencies();
     final Dag dag = new Dag(this.name, this.dagProcessor);
     final Map<NodeBuilder, Node> builderNodeMap = createBuilderToNodeMap(dag);
     updateNodesRelationships(builderNodeMap);
-    // todo HappyRay: circular dependency detection.
     return dag;
   }
 
+  /**
+   * Checks if the builder contains nodes that form a circular dependency ring.
+   *
+   * <p>The depth first algorithm is described in this article
+   * <a href="https://en.wikipedia.org/wiki/Topological_sorting">https://en.wikipedia.org/wiki/Topological_sorting</a>
+   * </p>
+   *
+   * @throws DagException if true
+   */
+  private void checkCircularDependencies() {
+    class CircularDependencyChecker {
+
+      // The nodes that need to be visited
+      private final Set<NodeBuilder> toVisit = new HashSet<>(DagBuilder.this.builders);
+
+      // The nodes that have finished traversing all their parent nodes
+      private final Set<NodeBuilder> finished = new HashSet<>();
+
+      // The nodes that are waiting for their parent nodes to finish visit.
+      private final Set<NodeBuilder> ongoing = new HashSet<>();
+
+      // One sample of nodes that form a circular dependency
+      private final List<NodeBuilder> sampleCircularNodes = new ArrayList<>();
```

Comment:
Is sampleCircularNodes referring to circular results given a DAG? Why is it called `sample` ?

Reply:
Because there may be others.